商品購入機能

unitoro0903 opened the pull request:
不具合解消のため

unitoro0903 committed:
turbo消す

不具合解消のため

unitoro0903 committed:
formオブジェクト導入

複数のテーブルの情報を取り扱うために、フォームオブジェクト作成

unitoro0903 committed:
商品詳細、編集、購入画面に売却済みの商品はうつれない機能

画面移れないようにするためにif文とsoldoutメソッドで設定

unitoro0903 committed:
クレジットカード決済機能

クレジットカード決済機能の実装のために、PAYJP導入
鍵はコードに書けないので、環境変数に設定

unitoro0903 committed:
テストコード

unitoro0903 commented:
- [ https://gyazo.com/0bfc6d5a3720d8013115de5bdadc1348] 必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
- [ https://gyazo.com/153aa988d09a7825ec463592016d52a1] 入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
- [https://gyazo.com/75da2b504a3c327b4c2e831246570b20 ] ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
- [https://gyazo.com/b3f4b00bc4246227d8e8a50783252429 ] ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
- [https://gyazo.com/05cc24b15094fda7a00326cfbc178ba0 ] ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
- [ https://gyazo.com/26305a01d827f2ddb0b5e4d37e1f0361] 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
- [https://gyazo.com/10303ff67c5268e3a812d66dc562df10 ] 売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
- [https://gyazo.com/019c48a407fc3f8c7bbb3fcf26a01fa6 ] ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
- [ https://gyazo.com/a6559721f457914d3d24e4a2f7c5d2fb] ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
- [ https://gyazo.com/ed02461db9a8a0ce4523cb2e82cd4b13] テスト結果の画像

yuri240 commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,60 @@
+
+  def index
+    gon.public_key = ENV["PAYJP_PUBLIC_KEY"]
+    @item = Item.find(params[:item_id])
```
`index`, `create`アクションに定義されている`@item = Item.find(params[:item_id])`は
同一ファイル内に記述が重複しているので、
private以下にメソッドを定義し、`before_action`で呼び出しましょう。

yuri240 commented on app/models/record_address.rb:
```diff
@@ -0,0 +1,29 @@
+    validates :region_id, presence: true
+    validates :city, presence: true
+    validates :house_number, presence: true
+    validates :phone_number, presence: true, length: { in: 10..11 }, format: { with: /\A[0-9]+\z/, message: "は半角数値のみ入力してください" }
```
12~17行目に記述されている`presence: true`は削除しましょう！

理由：
```
7行目の`presence: true`が適用されているためです。
```

yuri240 commented on app/views/items/index.html.erb:
```diff
@@ -129,34 +129,33 @@
+<% end %>
       <%# //商品のインスタンス変数になにか入っている場合、中身のすべてを展開できるようにしましょう %>
 
       <%# 商品がない場合は以下のダミー商品が表示されるようにしましょう %>
```
159,161行目のコメントアウトですが、
これらは実装案内用となり、実装後は不要な記述となります。

したがって、
実装が完了し、不要となれば削除しましょう。

【理由】
企業提出時に実装案内コメントが残らないようにするためです。

また、もし必要であれば、第三者でもわかるように
「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめ致します。

yuri240 commented on spec/factories/record_addresses.rb:
```diff
@@ -0,0 +1,14 @@
+    token {"tok_abcdefghijk00000000000000000"}
+
+    association :user
+    association :item
```
11,12行目の記述は不要となりますので、削除していただけると良さそうです。

理由：
```
`spec/models/record_address_spec.rb`内でitemとuserの紐付けを行なっているためです。
```

yuri240 commented on app/models/record_address.rb:
```diff
@@ -0,0 +1,29 @@
+    # validates :record_id
+    validates :token, presence: true
+    validates :postal_code, presence: true, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: "は「3桁ハイフン4桁」の半角文字列で入力してください" }
+    validates :region_id, presence: true
```
```ruby
validates :region_id, presence: true
```
アクティブハッシュのカラムは、「空の場合」＝「id: 1を選択した場合」保存できないとなるので、
`presence: true`の代わりに`numericality`を用いましょう！

yuri240 commented on spec/models/record_address_spec.rb:
```diff
@@ -0,0 +1,95 @@
+        expect(@record_address.errors.full_messages).to include("Postal code can't be blank")
+      end
+      it 'region_idが空だと保存できないこと' do
+        @record_address.region_id = ""
```
アクティブハッシュのデータが登録できないのは、
id: 1（'---'）を選択した時であるはずです。

したがって、右辺を`""`ではなく、`1`に変更しましょう！

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
コメントさせていただきましたので、ご確認お願いいたします。

```ruby
◎次回提出時のお願い3点◎

①プルリクエストのタイトルを「turbo消す」ではなく「商品購入機能」に変更しましょう。

②プルリクエストに**What**と**Why**を記述しましょう。
⇨What（`どのような実装をしているのか`）と Why（`なぜこの実装が必要なのか`）を書くことで、
このブランチは何を実装していて、なぜその実装を行ったのかが他の人に伝えることができます。

②テストコード修正後は、改めてテストコード実行後のキャプチャ画像を撮影いただき、
こちらのプルリクエストへ添付していただくようお願いいたします。
```

ご不明点や疑問点がありましたら、チャット・コネクトメンターにご相談ください。

unitoro0903 committed:
テストコード修正

 テスト結果の画像

https://gyazo.com/8f6a00a6ae9fe27bad7e7f164d04e941

空の場合より「id: ０を選択した場合」が今回ふさわしいため、numericalityを用い変更しました

unitoro0903 committed:
不要な記述消す

企業提出時に実装案内コメントが残らないようにするために不要な記述消しました

unitoro0903 committed:
重複部分をメソッド化

保守性、可読性のため、重複部分をメソッド化しました

unitoro0903 committed:
presence: trueの重複部分消す

7行目の`presence: true`が適用されているため、不要な presence: trueは消しました

アクティブハッシュのidを指定したいのでnumericality: に変更しました

kota0901 commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,63 @@
+
+  def new
+    @record_address = RecordAddress.new
+  end
```
newアクションは今回不要です。削除しましょう。

理由は以下の2点です。
・もし仮にviewファイルに`records/new.html.erb`があるならば必要だが今回はそれがいないため
・indexアクションと内容が重複しているため

kota0901 commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  @item = Item.find(params[:item_id])
+end
+
+# end
```
endの位置について見直しましょう。

・32行目のend→不要
・62行目のend→必要

32行目のendは、現状のコードですと「classを閉じるためのend」となっています。
32行目でclassを閉じてしまうと、33行目以降のprivateメソッドが RecordsControllerクラスとして認識されないため、予期せぬエラーが発生してしまう恐れがあるためです。

kota0901 commented on app/models/record.rb:
```diff
@@ -0,0 +1,7 @@
+
+  belongs_to :item
+  belongs_to :user
+  has_one :ship
```
5行目の記述について、配送先住所モデルとアソシエーションが組めているか確認しましょう。

kota0901 reviewed (commented):
レビュー依頼ありがとうございます！
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

unitoro0903 committed:
不必要なアクション削除

今回はnewアクションは必要ないため、削除
また、endも不必要なところは削除

unitoro0903 committed:
アソシエーション変更

配送先住所のモデル名を間違えていたので
正しいものに変更

Akinori0123 commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,61 @@
+def set_item
+  @item = Item.find(params[:item_id])
+end
+
```
・app/controllers/records_controller.rb
現在1行目のclassに対応するendが無いように見受けられますが、正常に購入はできますでしょうか？
もし構文エラーが発生する場合は前回レビュー通りに末尾にendを追加しましょう。

Akinori0123 commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,61 @@
+
+  def index
+    gon.public_key = ENV["PAYJP_PUBLIC_KEY"]
+    set_item
```
```
set_item
```
上記は削除しましょう。
4行目で既に呼び出されているためです。

Akinori0123 commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,61 @@
+  
+
+  def create
+    set_item
```
```
set_item
```
上記も4行目で既に呼び出されているため削除しましょう。

Akinori0123 commented on db/migrate/20240726103523_create_orders.rb:
```diff
@@ -0,0 +1,8 @@
+      t.timestamps
+    end
+  end
+end
```
・db/migrate/20240726103523_create_orders.rb
上記マイグレーションファイルはREADMEに記載のないテーブルを作るものになりますので、rollbackの上削除しましょう。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
前回の修正内容確認いたしました。修正内容と軽微なリファクタリング面についてコメントいたしましたのでご確認のほどよろしくお願いいたします。

unitoro0903 committed:
不要なメソッド呼び出し部分の削除とendの加筆

beforeアクションで呼び出しているので、不要な部分削除
足りていないend加筆

unitoro0903 committed:
不要なファイル削除

必要ないので削除

TakashiKozuma commented on app/controllers/records_controller.rb:
```diff
@@ -0,0 +1,61 @@
+end
+
+def redirect_if_owner_or_sold_out
+  @item = Item.find(params[:item_id])
```
```rb
@item = Item.find(params[:item_id])
```
set_itemと記述が重複しているので削除しましょう。
削除するだけだとエラーになるので、3,4行目の順番を入れ替えて先にset_itemが読み込まれるように修正すれば、エラーを回避できます。

TakashiKozuma commented on spec/models/record_address_spec.rb:
```diff
@@ -0,0 +1,95 @@
+
+RSpec.describe RecordAddress, type: :model do
+  let(:user) { FactoryBot.create(:user)}
+  let(:item) { FactoryBot.create(:item, user_id: user.id)}
```
```Rb
let(:item) { FactoryBot.create(:item, user_id: user.id)}
```
現状の記述の仕方だと、商品の出品者と購入者が同じユーザーになってしまうので、`, user_id: user.id`の部分は削除しましょう。

TakashiKozuma reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認致しました。
コメントを追記いたしました。
ご確認お願いいたします。

修正後はコミット・プッシュのし忘れがないかご確認の上、再度フォームからレビュー依頼をお願い致します。

unitoro0903 committed:
順序変更

エラー回避のため、beforeアクションの順序変更

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
コミット内容確認いたしました。LGTMです。
マージ後、mainブランチでpullしていただければと思います。

購入機能の実装お疲れ様でした。
引き続き学習頑張っていきましょう！

unitoro0903 committed:
不要な記述削除
